File: src/java/others/easy/dtwenty/BinaryTreePaths.java
package java.others.easy.dtwenty;

import java.others.tools.TreeNode;

import java.util.ArrayList;
import java.util.List;

/**
 * 257 Binary Tree Paths
 *
 * Given a binary tree, return all root-to-leaf paths.
 *
 * Note: A leaf is a node with no children.
 *
 * Example:
 *
 * Input:
 *
 *    1
 *  /   \
 * 2     3
 *  \
 *   5
 *
 * Output: ["1->2->5", "1->3"]
 *
 * Explanation: All root-to-leaf paths are: 1->2->5, 1->3
 */

public class BinaryTreePaths {

    private static List<String> binaryTreePaths(TreeNode root) {

        List<String> result = new ArrayList<>();

        if (root == null)
            return null;

        if (root.left == null && root.right == null)
            result.add("" + root.val);

        if (root.right != null) {
            for (String oneRode : binaryTreePaths(root.right))
                result.add(root.val + "->" + oneRode);
        }

        if (root.left != null) {
            for (String oneRode : binaryTreePaths(root.left))
                result.add(root.val + "->" + oneRode);
        }

        return result;
    }
}

File: leetcode-scala/src/dfs/easy/BinaryTreePaths.scala
package dfs.easy

import tools.TreeNode

/**
 * @package dfs
 * @author 余天堂
 * @create 2020/7/12 23:33
 * @description BinaryTreePaths
 */
object BinaryTreePaths {

    /**
     * 680 ms   20.00%
     * 51.4 MB  100.00%
     * @param root
     * @return
     */
    def binaryTreePaths(root: TreeNode): List[String] = {
        if (root == null) return Nil
        if (root.left == null && root.right == null) return List(root.value.toString)
        val paths: List[String] = binaryTreePaths(root.left) ::: binaryTreePaths(root.right)

        paths.map(x => root.value.toString + "->" + x)
    }

}
